WikiWatershed/model-my-watershed
# Prevent Empty AoIs

## Overview

For some time there has been a mysterious behavior that empties out the areas of interest in projects. We have been unable to pinpoint the cause of this. The work in this PR attempts at adding guards to prevent this from happening in the future.

The manner in which most people interact with projects is via the MMW UI, since project manipulation is not available in the public API. Given the size and complexity of MMW's UI, it is possible there are some edge cases which empty out the area of interest, which then gets saved, causing this. To prevent this, we add a validator to the Project Update Serializer, which is the only way an existing project is updated via the UI, to validate that the AoI is not empty.

In case this is happening via other means, we also add a database constraint to ensure that the area of interest cannot be NULL. This will ensure that no matter what updates the project: Django or direct SQL, this will always be respected. In order for this constraint to be applied, any existing projects with NULL areas of interest will have to be deleted. We will export those projects from production before deploying this update.

An exemption is made for Activities, which are Projects created via ITSI mode. These Projects, marked by `is_activity=True`, are allowed to have empty AoIs. As far as we know, these Activities were never used in production, but they did weaken the protections around Area of Interest in b81edefa, laying the ground for this bug.

Connects #3068 

## Testing Instructions

### Preparation

Before you check out this PR, ensure some behaviors on `develop`:

#### Empty AoI Project

* Open MMW on `develop` and log in
* Create a new project (TR-55 may be simpler, but this would apply to both TR-55 and GWLF-E projects)
* Open the network tab of developer tools
* Rename the project to be "Empty AoI"
* Find that PUT request in the network tab, and copy it as a `curl` command
* Paste that in a text editor, find `area_of_interest` key in the request body, and set its value to `null`
* Run that curl command with `area_of_interest: null` set
    - [x] Ensure the command runs successfully, and the project's AoI is deleted

#### Activity Project

* Create another new project, and note its id (referenced later as `ACTIVITY_PROJECT_ID`)
* In your terminal, run `./scripts/manage.sh shell_plus`
* In the Django shell, run the following:
    ```python
    a = Project.objects.get(pk=ACTIVITY_PROJECT_ID)
    a.is_activity = True
    a.area_of_interest = None
    a.save()
    ```

### Testing

* Check out this branch
* Run the `curl` command with the empty AoI again
  - [x] Ensure you get a 400 with a validation error
* Run migrations `./scripts/manage.sh migrate`
  - [x] Ensure the Empty AoI Project has been deleted
  - [x] Ensure the Activity Project is still there
* Run `shell_plus` and select any project other than the Activity Project
    ```python
    p = Project.objects.get(pk=NON_ACTIVITY_PROJECT_ID)
    ```
* Try and set its AoI to NULL
    ```python
    p.area_of_interest = None
    p.save()
    ```
    - [x] Ensure you get an error
* Select the activity project, and try to mark it as a non-activity
    ```python
    a = Project.objects.get(pk=ACTIVITY_PROJECT_ID)
    a.is_activity = False
    a.save()
    ```
    - [x] Ensure you get an error
